mautic/mautic
# Fix ampersand encoding in tokens

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [x]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [x] 
| Related user documentation PR URL      | N/A
| Related developer documentation PR URL | N/A
| Issue(s) addressed                     | Fixes #11643 



#### Description:



This PR will make sure decimal or hex encoded ampersands are properly handled. As you can see in #11643, ampersands in tokens are replaced with `&#38;`. This is caused by [`InputHelper::clean`](https://github.com/mautic/mautic/blob/4.4/app/bundles/CoreBundle/Helper/InputHelper.php#L193). I don't feel really comfortable in changing that logic, so I extended the current 'decode ampersands' functionality to support decimal and hex encoded ampersands.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester));
2. Make sure you have a contact;
3. Create a template email and use `{url}` as the URL of a link;
4. Call the API endpoint: `/api/emails/[ID OF EMAIL]/contact/[ID OF CONTACT]/send` with body:
```
{
    "tokens": {
        "url": "https://localhost/test?foo=bar&email=john.doe%40example.com"
    }
}
```
5. Click the link in the received email;
6. See you're redirected to `https://localhost/test?foo=bar&email=john.doe%40example.com` instead of `https://localhost/test?foo=bar&#38;email=john.doe%40example.com`.
